Zigbee Green Power Support

konistehrad opened the pull request:
Copied from the originating comment. Goals for this fork are as follows:

- Implement the minimum functionality to conform to being a Green Power Target as per the Green Power Specification
- Create a GreenPowerController class that will do state tracking as per Figures 67-68 of the ZGP spec
- Create ZHA-compatible Devices that are transparently usable by consuming code

Specifically, this means we do not directly receive or transmit Green Power frames. We ask Green Power Proxy devices on the network to do that. The code intends to be flexible enough to allow for that in the future to become a Green Power Target+ device, though that will involve work in all of the implementing `ControllerApplication`s and their respective NCPs.

Commit:
Mandatory attrs and commands for minimal GPT support

konistehrad committed:
Correct command tx/rx location, add 242 endpoint with clusters.

konistehrad committed:
Move cluster back to general

This preserves compatibility with consuming code. Start also on trying to
subscribe to events in the GPS state machine

konistehrad commented:
Currently I'm looking for feedback on the overall structure of the project, and also some help in figuring out the best way to parse incoming ZCL messages on the GreenPower clusters. Namely:

- What's the correct object to subscribe to to get this information? I know `Endpoint::handle_message` is deprecated, but I don't see a clear path forward for an alternative
- What method signatures will I need to consume this data? I haven't quite unpacked everything going on in `zigpy.listeners` yet.

Thanks in advance for any interest and help.

puddly commented:
> What's the correct object to subscribe to to get this information? I know Endpoint::handle_message is deprecated, but I don't see a clear path forward for an alternative

Zigpy's packet parsing is currently in a state of flux but I'm hoping to simplify it for use cases such as this. Currently, there is a callback/synchronous listener interface:

https://github.com/zigpy/zigpy/blob/4570a0a47b6f2ad56f922f8cd5861bbfdda25915/zigpy/application.py#L1059-L1106

You can get a feel for how I was hoping it to behave from the unit tests: https://github.com/zigpy/zigpy/blob/dev/tests/test_listeners.py

konistehrad committed:
Start sending commissioning commands around

Broadcast path is the old ControllerApplication.broadcast
path for the time being, which isn't ideal but also
having it all encapsulated in one class isn't the worst

konistehrad committed:
Fix removed GP cluster file in __init__

konistehrad committed:
Reorganize permit messages in green power controller

Hedda commented:
> I've started working in earnest on adding ZGP support to ZHA. My current branch is here: https://github.com/konistehrad/zigpy/tree/zgp
> 
> …
> 
> Currently I'm looking for feedback on the overall structure of the project, and also some help in figuring out the best way to parse incoming ZCL messages on the GreenPower clusters.

Awesome! Developers interested in helping with getting native Zigbee Green Power End Device support working in zigpy should be aware that several non-developers have previously offered to donate ZGP hardware devices to developers willing to work on ZGP support in zigpy, (e.g. batteryless and energy-harvesting switch/remote based on EnOcean PTM215ZE, PTM215Z, or PTM216Z modules, such as example "Friends of Hue" switches), so if you or other zigpy developers do not have any ZGP devices then you/they might want to consider contacting them to get some free hardware samples.

For reference; support for native Zigbee Green Power End Device in zigpy has previously also been discussed and tracked in this feature request issue -> https://github.com/zigpy/zigpy/issues/341 as well as in this previous pull request -> https://github.com/zigpy/zigpy/pull/656 (which in turn was a continuation of PR #519)

--------------------------------

**Note!** If you are a non-developer (e.g. only an end-user of the [Zigbee Home Automation integration in Home Assistant](https://www.home-assistant.io/integrations/zha)) then you should only click to just give a thumbs-up to the top-post and comments to show you like this, **but please do not post "+1" or "me too" and other non-constructive comments to this issue that is not directly related to the development of ZGP support in zigpy,**

- https://github.com/zigpy/zigpy/issues/341

As non-developers please consider only reading this and instead post comments to the Home Assistant's community feature request for general or off-topic discussions that are not directly constructive to developers for the development of ZGP support in the zigpy library, so see -> https://community.home-assistant.io/t/zha-integration-to-get-full-support-for-zgp-zigbee-green-power-devices/277149

<img width="744" height="215" alt="image" src="https://github.com/user-attachments/assets/54b75957-473c-47f6-b2ab-792789e6d6fc" />

konistehrad committed:
Remove yet another old reference

konistehrad committed:
Better broadcast logic compliments puddly

konistehrad commented:
Thanks @Hedda I've got a couple and pudd has been kind enough to offer more for testing. Thanks again for your help, interest and patience as we work to ensure this patch is robust enough for main

konistehrad committed:
Minor reorg. and fix logging format

konistehrad committed:
Do the thing I didn't wanna do and parse GDPFs

konistehrad committed:
Proper GPDF deserialization and Struct

konistehrad committed:
Get ready to treat direct GDPF's and commissioning

konistehrad committed:
calculate MICs

konistehrad committed:
fix server-to-client unicast command structure

konistehrad committed:
Turns out tunneled broadcast frames don't "come" from the proxy, they come from the device itself!

That's fine, let's just forward those packets from unknown ZGP profile devices into the
GreenPowerController and it can sort that all out. This should theoretically only happen during
commissioning anyway, so we test to see if we expect
weirdo broadcast packets before doing anything.

konistehrad committed:
Parse ZGP cluster options via structs

Also break it out into its own file but maintain the export for
backwards compatibility

konistehrad committed:
Perform broadcast group subscription

konistehrad committed:
blow up pairing options for now

works around an endianness issue in Struct deserialization

konistehrad committed:
Track commissioning window closure time

konistehrad committed:
Work on device commissioning with device creation

konistehrad committed:
Small fixes

konistehrad committed:
At long last I understand how Optional works

konistehrad committed:
Hey StructField is awesome actually

Break out schemas into structs with requires and optional falgs

konistehrad committed:
Fix data initialization order

konistehrad committed:
audit encryption paths

konistehrad committed:
Fix key protection via A.1.5.9.1

konistehrad committed:
Well the device shows up in ZHA anyway

konistehrad committed:
Give ZGP its own top level

konistehrad committed:
Send pairing commands properly

Doesn't work yet, but we're getting somewhere.

konistehrad committed:
Correct Pairing notification at last

konistehrad committed:
Use Notification structure from Wireshark

This doesn't seem to be in the spec, but these fields match up with the
data we're getting so...

konistehrad committed:
Use Direction type in GP cluster

konistehrad committed:
try to use device status

konistehrad committed:
Fix device serialization

konistehrad committed:
Move commands into commands file

konistehrad committed:
External data storage (temp)

konistehrad committed:
Audit ext db paths

konistehrad committed:
Endpoint cluster type hints, as a treat

konistehrad committed:
Save ZGP data to DB, start work on bespoke ZGP device testing

konistehrad committed:
Fix ZGP database interactions

konistehrad committed:
that simply should not have been there

konistehrad committed:
cluster handle_message in ZGP device

konistehrad committed:
whoops, this needs to be a commandSchema

konistehrad committed:
event signals work properly in ZHA

konistehrad committed:
Tuya Self-powered stuff

konistehrad committed:
Stub out PTM215ZE device in prep for arrival

konistehrad committed:
send a slimmed down notification type up to ZHA

prevents barfing on LVBytes

konistehrad committed:
I named the file wrong?

konistehrad committed:
Fix some names that were bad.

konistehrad committed:
Fix unicast commissioning mode

Also send notification response packets, tho it looks like Hue bulbs
don't support them, the spec calls for a notification_response
after a generated notification in unicast forwarding mode

konistehrad committed:
track GPD frame counters

konistehrad committed:
Stop sending notify responses as nobody likes it

konistehrad committed:
fix missing SinkTableEntry setter

konistehrad committed:
Confirm PTM215ZE security parameters

konistehrad committed:
Allow Application subclass to instantiate GP Controller

Allows for specific NCP GP controller implementations

konistehrad committed:
Subscribe endpoint to broadcast group

Fixes EZSP broadcast commissioning. This isn't required, for some radios, but the radios that do need it, er, need it.

konistehrad committed:
Rx-on commissioning path with encryption

konistehrad committed:
Update SWS200 commadns

konistehrad committed:
try to spruce up Rx-on codepaths

konistehrad committed:
I misunderstood what "encrypted" meant here.

konistehrad committed:
Bring comm-mode logic back

konistehrad committed:
as a matter of pride, utilize zgp_encrypt to encode key

konistehrad committed:
get cluster commands ready

konistehrad committed:
Fix attribute reads on GPP cluster

konistehrad committed:
Send pairing when new GPP's join for known broadcast GPD's

konistehrad committed:
Fix up GPS attributes and values

konistehrad committed:
Give the GP broadcast group a name

konistehrad committed:
Readme note about support

konistehrad committed:
Bring derived NWK alias in line with spec

konistehrad committed:
Send Device_Annce for incoming GPD's

Resolves address conflicts with derived NWK addresses

konistehrad committed:
missed a comma, broke the DB initialization script

konistehrad committed:
prep possibility to assign NWK alias explicitly

konistehrad committed:
Utilize the "unicast" flag in proxy comm mode options

konistehrad committed:
improve Rx path with concentrated channel search messages

konistehrad committed:
Fix missed reference in appdb.py

konistehrad committed:
Tighten up EP/cluster checks, unicast rx cap pair responses

konistehrad committed:
Smarter broadcast-or-not commands

konistehrad committed:
Fix type-check-only import in GPD

konistehrad committed:
Remove old, bad "device type" ref from ZHA profile

konistehrad committed:
GP attribute notif data types

konistehrad committed:
Move GP quirks to ZHA quirks

konistehrad committed:
Actually handle GP attr updates!

konistehrad committed:
Running the linter and the linter hates me!

Hedda commented:
@konistehrad is your Green power notification handler PR for ZHA in Home Assistant core not dependent on this PR for zigpy? 

If so then should that PR not also be left as a draft until this PR is merged (and zigpy bumped)? -> https://github.com/home-assistant/core/pull/106118

konistehrad commented:
@Hedda Good check: no, [the associated HA core patch](https://github.com/home-assistant/core/pull/106118) does not depend on this. I intentionally wrote that patch to allow independent and async review and deployment.

puddly committed:
Fix up lint errors

puddly committed:
Merge branch 'dev' into zgp

puddly committed:
Get pre-commit passing

puddly committed:
Ensure type annotations are 3.8-compatible

puddly committed:
Merge branch 'dev' into zgp

puddly committed:
Create `GreenPowerController` without a helper method

puddly committed:
Fix `warn` deprecation notice

puddly committed:
Get unit tests passing

puddly committed:
Fix 3.8 unit tests

konistehrad commented on zigpy/zgp/message_concentrator.py:
```diff
@@ -1,3 +1,5 @@
+from __future__ import annotations
```
Ah, sorry. The HA version I was testing against required 3.11, which I believe has annotations by default.

konistehrad committed:
Merge remote-tracking branch 'upstream/dev' into zgp

konistehrad committed:
Add type annotation to GreenPowerProxy.cluster_id

jottr commented:
Is this PR still being worked on, or is there another place where development to support Green Power devices is happening?
Where would contributions be welcome to get Green Power integrated into zigpy?

[EDIT] After further reading it seems that https://github.com/zigpy/zigpy/issues/341 is the tracking issue for this feature. Is this correct?

puddly commented:
This PR is still the one that will be merged at some point, it just needs to be finished. Unfortunately, I haven't had time to pick up where @konistehrad left off just yet.

konistehrad commented:
Is there anything I can help with? I'm pretty swamped right now but I might be able to hop in and polish some stuff up if necessary.

Hedda commented:
FYI, CSA has released Green Power Revision 1.1.2 specifications for Zigbee-based Energy-harvesting Technology:

https://csa-iot.org/newsroom/the-connectivity-standards-alliance-releases-green-power-1-1-2-for-zigbee-based-energy-harvesting-technology/

The latest revision of the Green Power standard, Revision 1.1.2 introduces additional features and scripts that are supposed to simplify testing and allow Green Power developers to use the same unified test harness for Green Power receiver products as used for the traditional Zigbee protocol.

Zigbee PRO Green Power feature specification Basic functionality set version 1.1.2 can now be downloaded from CSA:

https://csa-iot.org/developer-resource/specifications-download-request/

klodeckl commented:
Hi, I would be very happy to see this feature soon. :)
I just migrated (from Deconz) to ZHA and can’t use my Tap now (installed in a Gira switch wall mounted).

SirUli commented:
Still relevant - do not close

stelb commented:
I have switches waiting for this  😬

corcaioch commented:
Is this largely code complete, and just requires testing, or is there much further development required here?

potelux commented:
> This PR is still the one that will be merged at some point, it just needs to be finished. Unfortunately, I haven't had time to pick up where @konistehrad left off just yet.

Is there a bounty somewhere for this feature?  Or is there another form of supporting your work?  I know it is hard to find time to work on these things, so I would gladly add to the incentive by supporting a bounty for its completion.

marvelloard commented:
Is someone working on this?

afsy-d commented:
No still active

ArchaicRider commented:
> I have switches waiting for this 😬

Me too!

marvelloard commented:
> > I have switches waiting for this 😬
> 
> Me too!

same

klodeckl commented:
I also had switches waiting for it. But meantime I replaced them, this takes too long. :(

konistehrad commented:
I may find myself with a little time in the near future. If there are changes I can make to the PR to get it ready for production I can take a checklist. I'll start by rebasing on main of course.

marvelloard commented:
> I may find myself with a little time in the near future. If there are changes I can make to the PR to get it ready for production I can take a checklist. I'll start by rebasing on main of course.

that would be incredible!

Hedda commented:
> > > _I have switches waiting for this 😬_
> > 
> > 
> > _Me too!_
> 
> _same_

@ArchaicRider @marvelloard @klodeckl and others, As mentioned multiple times in the matching ZGP feature request, PLEASE!

If you are a non-developer (e.g. a non-developing end-user of the [Zigbee Home Automation integration in Home Assistant](https://www.home-assistant.io/integrations/zha)) then you should only click to just give a thumbs-up to the top-post and comments to simply show you like this

<img width="744" height="168" alt="image" src="https://github.com/user-attachments/assets/523790ee-5572-4286-84a5-295161167f3b" />

**But please, do not post "_+1_" or "_me too_" and other non-constructive comments to this issue that is not directly related to the development of ZGP support in zigpy**, (as such posts send annoying spam everyone subscribing to this and do not contribute anything to the development).

- https://github.com/zigpy/zigpy/issues/341

As non-developers what you can do is to consider only following/reading this and instead post any end-user-pespective comments to the related Home Assistant's community feature request thread for general or off-topic discussions that are not directly constructive to developers for the development of ZGP support in the zigpy library, so please comment there instead:

- https://community.home-assistant.io/t/zha-integration-to-get-full-support-for-zgp-zigbee-green-power-devices/277149

konistehrad commented:
@puddly Given how long it's been, would it be preferable to split this massive PR into more piecemeal parts as I work to rebase on main? For example, I have a draft of the ZGP clusters and base types ready to go, cleaned up to better match the style of the rest of the project, but without the GreenPowerController or logic. Might be easier to get this feature landed if we're not looking at several thousand lines.

puddly commented:
Yes please! It's been a little while since I looked at ZGP but from what I recall this PR targets all ZGP device types, including ones using more secure encryption, correct? We can probably simplify things by splitting it into stages, supporting basic devices first and then progress from there.

It's a little difficult to review all at once in its current state because 3k lines of code without tests is just not something we can merge. We aim to have 100% test coverage, especially for all new code.

konistehrad commented:
OK great. I'll close this out tomorrow when I post up the cluster PR.

konistehrad commented:
New day, new PR. Thanks everyone for your patience!

ulope commented:
New PR is #1659 AFAICS?

konistehrad commented:
Yes though it won’t be the hulking monster this PR was. I’m going to do it piecemeal, so there will be several coming down.
